### PR TITLE
PHP 8: Refactoring `Notifications`

### PR DESCRIPTION
**Pull request by iszmais**
This refactors the `Notifications` to be PHP compatible with PHP 8 and 7.4 
further...
- Docs where updated
- A testsuite was added
- trivial code was removed
- multi class files where split
- a repository was implemented
- a namespace was introduced

### CODE REVIEW

File: Services/Awareness/classes/class.WidgetManager.php [hunk 1]
```diff
@@ -99,14 +100,14 @@ public function notifyOnNewOnlineContacts() : void
             return;
         }
 
-        $notification = new \ilNotificationConfig('who_is_online');
+        $notification = new ilNotificationConfig('who_is_online');
```
Comment (alex40724) [inline Services/Awareness/classes/class.WidgetManager.php:103]: A suggestion: It would be great to get this instance from a factory that is somehow available through the DIC.

Comment (mjansenDatabay) [reply Services/Awareness/classes/class.WidgetManager.php:103]: Good point @alex40724

File: Services/Object/classes/class.ilObjectFactory.php [hunk 2]
```diff
@@ -246,6 +246,8 @@ protected function includeClassIfNotExists(
         string $a_obj_type,
         ilObjectDefinition $objDefinition
     ) {
+        require_once("libs/composer/vendor/autoload.php");
```
Comment (mjansenDatabay) [inline Services/Object/classes/class.ilObjectFactory.php]: Please remove this.

File: Services/UICore/classes/class.ilGlobalPageTemplate.php [hunk 3]
```diff
@@ -5,6 +5,7 @@
 use ILIAS\HTTP\Services as HTTPServices;
 use ILIAS\DI\UIServices;
 use ILIAS\GlobalScreen\Services;
+use ILIAS\Notifications\ilNotificationOSDGUI;
```
Comment (mjansenDatabay) [inline Services/UICore/classes/class.ilGlobalPageTemplate.php:8]: Why is this needed?

File: src/GlobalScreen/Scope/MetaBar/Factory/NotificationCenter.php [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 use ILIAS\GlobalScreen\Identification\IdentificationInterface;
 use ILIAS\GlobalScreen\Scope\MetaBar\Collector\Renderer\NotificationCenterRenderer;
+use ILIAS\GlobalScreen\Scope\Notification\Factory\isItem as notaIsItem;
```
Comment (mjansenDatabay) [inline src/GlobalScreen/Scope/MetaBar/Factory/NotificationCenter.php]: Why did you introduce this alias?

File: src/GlobalScreen/Scope/MetaBar/Factory/NotificationCenter.php [hunk 5]
```diff
@@ -21,7 +22,7 @@
  * Class NotificationCenter
  * @author Fabian Schmid <fs@studer-raimann.ch>
  */
-class NotificationCenter extends AbstractBaseItem implements isItem, hasSymbol
+class NotificationCenter extends AbstractBaseItem implements hasSymbol
```
Comment (mjansenDatabay) [inline src/GlobalScreen/Scope/MetaBar/Factory/NotificationCenter.php:25]: Why is this removed?

**Comment by mjansenDatabay** [comment]
@Uwe-Kohnle FYI: These are the changes you will have to "PHP 8-review" after this PR has been merged.

File: src/GlobalScreen/Scope/MetaBar/Factory/NotificationCenter.php (hunk 5 above)
Comment (iszmais) [reply src/GlobalScreen/Scope/MetaBar/Factory/NotificationCenter.php:25]: cause this is already required in AbstractBaseItem.
Im aware that this is no argument to remove a requirement but since the NotificationCenter has no other depencdency to item except trhough teh abstract base item this requirement should stay and fall with it.

File: src/GlobalScreen/Scope/MetaBar/Factory/NotificationCenter.php (hunk 4 above)
Comment (iszmais) [reply src/GlobalScreen/Scope/MetaBar/Factory/NotificationCenter.php]: To prevent mistakings with the isItem in the same directory. This is another item

File: Services/UICore/classes/class.ilGlobalPageTemplate.php (hunk 3 above)
Comment (iszmais) [reply Services/UICore/classes/class.ilGlobalPageTemplate.php:8]: I introduced namespaces within this refactoring. Therefore the few external accessors (i assume its just this one) new to use this namespace.

File: Services/Awareness/classes/class.WidgetManager.php (hunk 1 above)
Comment (iszmais) [reply Services/Awareness/classes/class.WidgetManager.php:103]: I agree. But i assume this would be better fitting in a own PR.

File: Services/Object/classes/class.ilObjectFactory.php (hunk 2 above)
Comment (iszmais) [reply Services/Object/classes/class.ilObjectFactory.php]: Of course. This was for testing. removed it.

File: Services/Notifications/classes/ilNotificationOSDRepository.php [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+/**
+ * @author Jan Posselt <jposselt@databay.de>
+ */
+class ilNotificationOSDRepository
+{
+    private static ?self $instance = null;
+    private ilDBInterface $database;
+
+    public static function getInstance() : self
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/ilNotificationOSDRepository.php:37]: Do we really need to introduce a singleton here?

Maybe you can add something like `NotificationServices` to the `$DIC` and add a method there to retrieve a repository?

Comment (iszmais) [reply Services/Notifications/classes/ilNotificationOSDRepository.php:37]: I Agree, but would prefer todo so in a later PR since this dont seem to be relevant to be PHP8 usable and is more of a general refactoring.

**Comment by iszmais** [comment]
Hi @alex40724 and @mjansenDatabay ,
This is now finally ready for a pre-review. Feel free to leave me a response or to merge it.

Greetings,
@iszmais

File: src/GlobalScreen/Scope/MetaBar/Factory/NotificationCenter.php (hunk 4 above)
Comment (mjansenDatabay) [reply src/GlobalScreen/Scope/MetaBar/Factory/NotificationCenter.php]: Then I suggest to call it `IsNotificationItem`

File: Services/Notifications/test/notifications/ilNotifcationOSDTest.php [hunk 7]
```diff
@@ -0,0 +1,136 @@
+<?php declare(strict_types=1);
+
+use PHPUnit\Framework\TestCase;
```
Comment (mjansenDatabay) [inline Services/Notifications/test/notifications/ilNotifcationOSDTest.php]: Please put  the license header on top

File: Services/Notifications/classes/Model/ilNotificationConfig.php [hunk 8]
```diff
@@ -0,0 +1,281 @@
+<?php declare(strict_types=1);
+
+namespace ILIAS\Notifications\Model;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/Model/ilNotificationConfig.php]: Please put  the license header on top

File: Services/Notifications/classes/Model/ilNotificationLink.php [hunk 9]
```diff
@@ -0,0 +1,67 @@
+<?php declare(strict_types=1);
+
+namespace ILIAS\Notifications\Model;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/Model/ilNotificationLink.php]: Please put  the license header on top

File: Services/Notifications/classes/Model/ilNotificationObject.php [hunk 10]
```diff
@@ -0,0 +1,58 @@
+<?php declare(strict_types=1);
+
+namespace ILIAS\Notifications\Model;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/Model/ilNotificationObject.php]: Please put  the license header on top

File: Services/Notifications/classes/Model/ilNotificationParameter.php [hunk 11]
```diff
@@ -0,0 +1,64 @@
+<?php declare(strict_types=1);
+
+namespace ILIAS\Notifications\Model;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/Model/ilNotificationParameter.php]: Please put  the license header on top

File: Services/Notifications/classes/Provider/NotificationCenterProvider.php [hunk 12]
```diff
@@ -1,11 +1,28 @@
-<?php namespace ILIAS\Notifications\Provider;
+<?php declare(strict_types=1);
+
+namespace ILIAS\Notifications\Provider;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/Provider/NotificationCenterProvider.php]: Please put  the license header on top

File: Services/Notifications/classes/Setup/ilNotificationSetupHelper.php [hunk 13]
```diff
@@ -1,21 +1,38 @@
-<?php
-
-require_once 'Services/Notifications/classes/class.ilNotificationDatabaseHelper.php';
+<?php declare(strict_types=1);
+
+namespace ILIAS\Notifications;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/Setup/ilNotificationSetupHelper.php:3]: Please put  the license header on top

File: Services/Notifications/classes/class.ilNotificationGUI.php [hunk 14]
```diff
@@ -1,52 +1,59 @@
-<?php
-/* Copyright (c) 1998-2010 ILIAS open source, Extended GPL, see docs/LICENSE */
+<?php declare(strict_types=1);
 
+use ilCheckboxInputGUI;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/class.ilNotificationGUI.php:20]: Please put  the license header on top

File: Services/Notifications/classes/class.ilObjNotificationAdmin.php [hunk 15]
```diff
@@ -1,9 +1,26 @@
 <?php declare(strict_types=1);
-/* Copyright (c) 1998-2021 ILIAS open source, Extended GPL, see docs/LICENSE */
 
-/**
- * Class ilObjNotificationAdmin
+use Exception;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/class.ilObjNotificationAdmin.php]: Please put  the license header on top

File: Services/Notifications/classes/class.ilObjNotificationAdminAccess.php [hunk 16]
```diff
@@ -1,9 +1,25 @@
 <?php declare(strict_types=1);
-/* Copyright (c) 1998-2009 ILIAS open source, Extended GPL, see docs/LICENSE */
 
-/**
- * Class ilObjNotificationAdminAccess
+use ilObjectAccess;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/class.ilObjNotificationAdminAccess.php]: Please put  the license header on top

File: Services/Notifications/classes/class.ilObjNotificationAdminGUI.php [hunk 17]
```diff
@@ -1,12 +1,30 @@
 <?php declare(strict_types=1);
-/* Copyright (c) 1998-2010 ILIAS open source, Extended GPL, see docs/LICENSE */
 
+use ilCtrlException;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/class.ilObjNotificationAdminGUI.php]: Please put  the license header on top

File: Services/Notifications/classes/ilNotificationEchoHandler.php [hunk 18]
```diff
@@ -0,0 +1,33 @@
+<?php declare(strict_types=1);
+
+namespace ILIAS\Notifications;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/ilNotificationEchoHandler.php]: Please put  the license header on top

File: Services/Notifications/classes/ilNotificationHandler.php [hunk 19]
```diff
@@ -0,0 +1,34 @@
+<?php declare(strict_types=1);
+
+namespace ILIAS\Notifications;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/ilNotificationHandler.php]: Please put  the license header on top

File: Services/Notifications/classes/ilNotificationMailHandler.php [hunk 20]
```diff
@@ -0,0 +1,44 @@
+<?php declare(strict_types=1);
+
+namespace ILIAS\Notifications;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/ilNotificationMailHandler.php]: Please put  the license header on top

File: Services/Notifications/classes/ilNotificationOSDGUI.php [hunk 21]
```diff
@@ -0,0 +1,78 @@
+<?php declare(strict_types=1);
+
+namespace ILIAS\Notifications;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/ilNotificationOSDGUI.php]: Please put  the license header on top

File: Services/Notifications/classes/ilNotificationOSDHandler.php [hunk 22]
```diff
@@ -0,0 +1,78 @@
+<?php declare(strict_types=1);
+
+namespace ILIAS\Notifications;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/ilNotificationOSDHandler.php]: Please put  the license header on top

File: Services/Notifications/classes/ilNotificationOSDRepository.php [hunk 23]
```diff
@@ -0,0 +1,122 @@
+<?php declare(strict_types=1);
+
+namespace ILIAS\Notifications;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/ilNotificationOSDRepository.php]: Please put  the license header on top

File: Services/Notifications/classes/ilNotificationSettingsTable.php [hunk 24]
```diff
@@ -1,25 +1,47 @@
-<?php
-
-require_once 'Services/Table/classes/class.ilTable2GUI.php';
-
+<?php declare(strict_types=1);
+
+namespace ILIAS\Notifications;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/ilNotificationSettingsTable.php]: Please put  the license header on top

File: Services/Notifications/classes/ilNotificationSystem.php [hunk 25]
```diff
@@ -0,0 +1,187 @@
+<?php declare(strict_types=1);
+
+namespace ILIAS\Notifications;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/ilNotificationSystem.php]: Please put  the license header on top

File: Services/Notifications/classes/ilNotificationUserIterator.php [hunk 26]
```diff
@@ -0,0 +1,81 @@
+<?php declare(strict_types=1);
+
+namespace ILIAS\Notifications;
```
Comment (mjansenDatabay) [inline Services/Notifications/classes/ilNotificationUserIterator.php]: Please put  the license header on top

File: Services/Notifications/test/bootstrap.php [hunk 27]
```diff
@@ -0,0 +1,4 @@
+<?php declare(strict_types=1);
+/* Copyright (c) 1998-2021 ILIAS open source, Extended GPL, see docs/LICENSE */
```
Comment (mjansenDatabay) [inline Services/Notifications/test/bootstrap.php]: Please use the new license header (exactly here)

File: Services/Notifications/test/ilNotificationsBaseTest.php [hunk 28]
```diff
@@ -0,0 +1,27 @@
+<?php declare(strict_types=1);
+
+use PHPUnit\Framework\TestCase;
```
Comment (mjansenDatabay) [inline Services/Notifications/test/ilNotificationsBaseTest.php]: Please put  the license header on top

File: Services/Notifications/test/ilServicesNotificationsSuite.php [hunk 29]
```diff
@@ -0,0 +1,69 @@
+<?php declare(strict_types=1);
+
+use PHPUnit\Framework\TestSuite;
```
Comment (mjansenDatabay) [inline Services/Notifications/test/ilServicesNotificationsSuite.php]: Please put  the license header on top

**Review by mjansenDatabay — CHANGES_REQUESTED** [review]
Hi @iszmais,

I left some remarks (almost all of them in regards to the license header). If my remarks are fixed, this PR can be merged.

Please address the following issues in a further PR:

- [x] We should get rid of the singleton
- [x] Please check which services of this component could be moved to some kind of `NotificationServices` container provided by the `$DIC` (let's have a chat)
- [x] Use the `RepositoryPattern` as documented
- [ ] Provide the `Client-side only Toast` in cooperation with @alex40724 for the (AFAIR) undocumented usage of the `Old OSD Notifications` in the `CoPage` context
- [x] Use `ilCtrlInterface` instead of `ilCtrl` in the GUI classes of this coponent
- [x] ... and maybe other things on your existing TODO list :)

Best regards,
@mjansenDatabay

**Review by mjansenDatabay — APPROVED** [review]
